codegen/facelift/facelift-codegen.py

```diff
@@ -355,18 +370,29 @@ def run_generation(input, output, dependency, libraryName, all):
                                                      for p in parameters))
     generator.destination = output
 
+    with open('{}/no_moc.cmake'.format(output), 'a+') as inFile:
+        inFile.write('set(HEADERS_NO_MOC_GENERATED\n')
+
     ctx = {'output': output}
     for module in system.modules:
         if module.name in modulesToGenerate:
+
             ctx.update({'module': module})
             module_path = '/'.join(module.name_parts)
             log.debug('process module %s' % module.module_name)
             ctx.update({'path': module_path})
+
             generateFile(generator, 'module/{{path}}/ModulePrivate.h', 'ModulePrivate.template.h', ctx, libraryName, "")
+            addToNoMocIfContainsMacro('module/{}/ModulePrivate.h'.format(module_path), output, noMocPath)
```

Comment:
at this point you can introspect local data and tell if there is MOC or no MOC. It will be faster than grepping file.

Reply:
Not sure what "introspect" means here, and what "local data" means.
The "generateFile" function can use some kind of grep on the template file, in order to detect if the generated file needs to be MOCed or not. The result of that grep can be cached so that no further grep will be needed when generating a file based on the same template.
Not sure whether this is the same idea as @bitmouse had.

Reply:
@bitmouse has meant and yes - as I've just figured out - we can inspect the model's "size" and add the file to the NO_MOC list if there are no any structures in the model.

Reply:
Grep is slow, we are going to avoid it whenever possible. If we can use something that works in constant time O(1), or even O(log n) then we will prefer that over grep, which is O(n). EOT.

Reply:
Here is how "slow" grep is:
`
$ time grep 'Q_OBJECT\|Q_ENUM\|Q_GADGET' ~/facelift/codegen/facelift/templates/*
/home/jacques/facelift/codegen/facelift/templates/Enum.template.h:    Q_GADGET
/home/jacques/facelift/codegen/facelift/templates/Enum.template.h:    Q_ENUM(Type)
/home/jacques/facelift/codegen/facelift/templates/ImplementationBaseQML.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/ImplementationBaseQML.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/ImplementationBase.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/IPCAdapter.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/IPCDBusServiceAdapter.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/IPCProxyAdapter.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/IPCProxy.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/IPCProxy.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/IPCServiceAdapter.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/ModulePrivate.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/QMLAdapter.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/ServiceMonitor.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/ServiceNotAvailableImpl.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/Service.template.h:    Q_GADGET
/home/jacques/facelift/codegen/facelift/templates/Service.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/ServiceWrapper.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/StructQObjectWrapper.template.h:    Q_OBJECT
/home/jacques/facelift/codegen/facelift/templates/Struct.template.h:    Q_GADGET

real	0m0,002s
user	0m0,001s
sys	0m0,001s
`

Reply:
Sorry, but this answer must be a joke. If you are trying to say that O(n) scales as good as O(1) this means that you completely do not understand it.
Now multiply that x1000. And 100k. And 1M. You can easily calculate this calculate that. More precisely grep will do O(k*n) where k is number of files and n number of bytes in the file. Checking the structure/list size on the generator side works in O(k) time, because we do it only once per file. 
I can only recommend 'Algorithms' by Robert Sedgewick and Kevin Wayne or another book 'Introduction to algorithms' by Thomas Cormen.

Reply:
What about reading my comments, before providing such irrelevant answers ??

Reply:
In which part my answer is irrelevant?